src/database/models.py

```diff
@@ -4379,5 +4381,180 @@ def simple_json(self):
+    slug = models.CharField(max_length=SHORT_STR_LEN, blank=True, null=True, unique=True)
+    latest_version = models.ForeignKey("CustomPageVersion", on_delete=models.SET_NULL, blank=True, null=True, related_name="latest_version")
+
+    def __str__(self):
+        return f"{self.title}"
+    
+    def save(self, *args, **kwargs):
+        if not self.slug:
+            self.slug = slugify(self.title)
+        super(CustomPage, self).save(*args, **kwargs)
+
+
+    def create_version(self):
+        version = CustomPageVersion(custom_page=self)
+        version.content = [x.simple_json() for x in self.content.all()]
+        version.save()
+
+        self.latest_version = version
+        self.is_published = True
+        self.save()
+        return version
+
+    def simple_json(self):
+        res = model_to_dict(self)
+        res["user"] = get_summary_info(self.user)
+        res["content"] = [e.simple_json() for e in self.content.all()]
+        return res
+
+    def full_json(self):
+        return self.simple_json()
+
+    class Meta:
+        db_table = "custom_pages"
+        ordering = ("title",)
+
+
+class CustomPageVersion(BaseModel):
+    """
+        A class used to represent a custom page version on the MassEnergize platform.
+        When a request is made to publish a page, a new version is created with the json blob of the page's 
+        elements. This version will be served to users when they visit the page. Admins can edit the page and create a new version.
+    """
+
+    custom_page = models.ForeignKey(CustomPage, on_delete=models.CASCADE, related_name="versions")
+    version = models.CharField(max_length=SHORT_STR_LEN, blank=True)
+    content = models.JSONField(blank=True, null=True, default=dict)
+    is_deleted = models.BooleanField(default=False, blank=True)
+    created_at = models.DateTimeField(auto_now_add=True)
+
+    def __str__(self):
+        return f"{self.custom_page.title} - {self.version}"
+    
+
+    def save(self, *args, **kwargs):
+        if not self.version:
+            now = datetime.datetime.now()
+            self.version = f"v{now.strftime('%Y%m%d%H%M%S')}"
+        super(CustomPageVersion, self).save(*args, **kwargs)
+    
+    def simple_json(self):
+        res = model_to_dict(self)
+        return res
+
+    def full_json(self):
+        return self.simple_json()
+
+    class Meta:
+        db_table = "custom_page_versions"
+        ordering = ("created_at",)
+
+
+
+class CustomPageBlock(BaseModel):
+    """
+        A class used to represent a custom page block(element) of a custom page on the MassEnergize platform
+        Blocks are the building blocks of a custom page. They can be of different types like text, image, video, section etc.
+    """
+    custom_page = models.ForeignKey(CustomPage, on_delete=models.CASCADE, db_index=True, related_name="content")
+    order = models.IntegerField(default=0, blank=True)
+    element = models.JSONField(blank=True, null=True, default=dict)
```

Comment:
@abdullai-t I figured the consensus was that `Elements` could have child `Elements` and the props would rather be stored as `JSONField` values ?